contracts/utils/UniversalResolver.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "../registry/ENS.sol";
+import "../resolvers/profiles/IExtendedResolver.sol";
+import "../resolvers/Resolver.sol";
+
+error OffchainLookup(address sender, string[] urls, bytes callData, bytes4 callbackFunction, bytes extraData);
+
+/**
+ * The Universal Resolver is a contract that handles the work of resolving a name entirely onchain,
+ * making it possible to make a single smart contract call to resolve an ENS name. 
+ */
+contract UniversalResolver is IExtendedResolver, ERC165 {
+    using Address for address;
+
+    ENS public immutable registry;
+
+    constructor(address _registry) {
+        registry = ENS(_registry);
+    }
+
+    function resolve(bytes calldata name, bytes memory data) external override view returns(bytes memory) {
+        Resolver resolver = findResolver(name);
+        if(address(resolver) == address(0)) {
+            return "";
+        }
+
+        if(resolver.supportsInterface(type(IExtendedResolver).interfaceId)) {
+            return callWithOffchainLookupPropagation(
+                address(resolver),
+                abi.encodeWithSelector(IExtendedResolver.resolve.selector, name, data),
+                UniversalResolver.resolveCallback.selector
+            );
+        } else {
+            return callWithOffchainLookupPropagation(
+                address(resolver),
+                data,
+                UniversalResolver.resolveCallback.selector
+            );
+        }
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns(bool) {
+        return interfaceId == type(IExtendedResolver).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    function callWithOffchainLookupPropagation(address target, bytes memory data, bytes4 callbackFunction) internal view returns(bytes memory ret) {
+        uint256 result;
+        uint256 size;
+        
+        assembly {
```

Comment:
For the assembly code what do you think about encapsulating inside their own functions to make it easier to understand what they do